include/TPP/IR/StructuredOpMatcher.h

```diff
@@ -0,0 +1,191 @@
+  Operand() = delete;
+  Operand(size_t idx) : idx(idx){};
+
+  const size_t idx;
+};
+
+struct IsProjectedPermutation {
+  IsProjectedPermutation() = default;
+  bool operator()(AffineMap map) const {
+    return map.isProjectedPermutation(/*allowZeroInResults=*/true);
+  }
+};
+
+struct IsIdentity {
+  IsIdentity() = default;
+  bool operator()(AffineMap map) const { return map.isIdentity(); }
+};
+
+struct HasStaticShape {};
+
+struct NumEqualsTo {
+  NumEqualsTo() = delete;
+  explicit NumEqualsTo(size_t value) : value(value){};
+  const size_t value;
+
+  bool operator()(size_t value) { return value == this->value; }
+};
+
+struct LessThanOrEqualTo {
+  LessThanOrEqualTo() = delete;
+  explicit LessThanOrEqualTo(size_t value) : value(value){};
+  const size_t value;
+
+  bool operator()(size_t value) { return value <= this->value; }
+};
+
+struct GreaterThanOrEqualTo {
+  GreaterThanOrEqualTo() = delete;
+  explicit GreaterThanOrEqualTo(size_t value) : value(value){};
+  const size_t value;
+
+  bool operator()(size_t value) const { return value >= this->value; };
+};
+
+struct HasAffineMapEqualsTo {
+  HasAffineMapEqualsTo() = delete;
+  explicit HasAffineMapEqualsTo(AffineMap map) : map(map){};
+  AffineMap map;
+
+  bool operator()(AffineMap map) const { return map == this->map; }
+};
+
+// OR or AND between predicates (functors).
+struct BinaryPredicate {
+public:
+  enum BinaryPredicateKind { _OR, _AND };
+
+private:
+  const BinaryPredicateKind binaryPredicateKind;
+
+public:
+  std::function<bool(size_t)> predicateOnLhs;
+  std::function<bool(size_t)> predicateOnRhs;
+
+public:
+  BinaryPredicateKind getKind() const { return binaryPredicateKind; }
+  BinaryPredicate() = delete;
+  BinaryPredicate(BinaryPredicateKind binaryPredicateKind,
+                  std::function<bool(size_t)> lhs,
+                  std::function<bool(size_t)> rhs)
+      : binaryPredicateKind(binaryPredicateKind),
+        predicateOnLhs(std::move(lhs)), predicateOnRhs(std::move(rhs)) {}
+};
+
+struct _OR : public BinaryPredicate {
+  _OR(std::function<bool(size_t)> lhs, std::function<bool(size_t)> rhs)
+      : BinaryPredicate(BinaryPredicateKind::_OR, lhs, rhs) {}
+
+  static bool classof(const BinaryPredicate *binaryPredicate) {
+    return binaryPredicate->getKind() == BinaryPredicateKind::_OR;
+  }
+};
+
+class StructuredOpMatcher {
+  using PredicateFn = std::function<bool(linalg::LinalgOp)>;
+
+public:
+  StructuredOpMatcher() = default;
+
+  StructuredOpMatcher(PredicateFn &&firstPredicate) {
+    predicates.push_back(std::move(firstPredicate));
+  }
+
+  template <typename OpTy> static StructuredOpMatcher make() {
+    return StructuredOpMatcher(
+        [](linalg::LinalgOp op) { return isa<OpTy>(op.getOperation()); });
+  }
+
+  bool match(Operation *op);
+
+  // Predicates on operation.
+  StructuredOpMatcher &hasBufferSemantics();
+  StructuredOpMatcher &hasTensorSemantics();
+  StructuredOpMatcher &inputs(std::function<bool(size_t)>);
```

Comment:
From interface, it is unclear how it differs from `input(AllOperands, ...)`.
Also, I would expect `inputs` to run my boolean predicate on all inputs (so Values? Operands?) but now it just passes `numDpsInputs`.

If it is really all it should do, then maybe rename it to `numInputs`.

Reply:
for me `input` was an op specific predicates while `inputs` for operand predicates. I can rename to `numDpsInputs`

Reply:
> Same extends to `BinaryPredicate` that is really specific right now. Maybe that's all we need but it looks like a design smell.

This is what I would like to have:
- predicate on operation. In this case the callable object can take an operation as input. Use case I have is: hasTensor/Buffer semantics or the number of Dps inputs/inits.
- predicate on the operands. The callable object takes only an op operand, here the use case I have at the moment are: query for an affine map or query for the type. Ideally you would pass an OpOperand but for how linalg allow you to query these properties we use an std::function<bool(AffineMap)> or HasStaticShape.
- predicate on the iterators: use cases: check iterators or number of dimensions
- predicate on the region, use cases: check the body of the generic.

Reply:
The only use case I have for BinaryPredicate is to check the number of input/output, agree that is a bit specific.

Reply:
I think I'd give whole linalg op for these predicates that care about `AffineMap`.
If we need an option for it to apply only to a single operand then I'd template the predicate `template<int idx> IsIdentity(linalgOp)` and call as `.operation(IsIdentity<2>()).